// @zoralabs/nft-components/dist/nft-full/BidHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BidHistory = void 0;
const react_1 = require("react");
const jsx_runtime_1 = require("react/jsx-runtime");
const nft_hooks_1 = require("@zoralabs/nft-hooks");
const react_2 = require("react");
const AddressView_1 = require("../components/AddressView");
const NFTDataContext_1 = require("../context/NFTDataContext");
const useMediaContext_1 = require("../context/useMediaContext");
const InfoContainer_1 = require("./InfoContainer");
const { format } = new Intl.NumberFormat();
const formatDate = (timestamp) => new Date(parseInt(timestamp, 10) * 1000).toLocaleString("en-US", {
    month: "long",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    hour12: true,
});
const BidHistory = ({ showPerpetual = true }) => {
    const { nft } = react_2.useContext(NFTDataContext_1.NFTDataContext);
    const { getString, getStyles } = useMediaContext_1.useMediaContext();
    const getPricingString = (pricing) => (jsx_runtime_1.jsxs("span", Object.assign({}, getStyles("pricingAmount"), { children: [format(parseFloat(pricing.prettyAmount)), " ", pricing.currency.symbol] }), void 0));
    const getPastBids = () => {
        var _a, _b, _c, _d, _e;
        const { data } = nft;
        if (!data) {
            return jsx_runtime_1.jsx(react_2.Fragment, {}, void 0);
        }
        const currentBid = ((_a = data.pricing.reserve) === null || _a === void 0 ? void 0 : _a.currentBid)
            ? [(_b = data.pricing.reserve) === null || _b === void 0 ? void 0 : _b.currentBid]
            : [];
        const eventsList = [
            ...(showPerpetual ? data.pricing.perpetual.bids : []),
            ...(((_c = data.pricing.reserve) === null || _c === void 0 ? void 0 : _c.previousBids) || []),
            ...currentBid,
        ].map((bid) => ({
            activityDescription: getString("BID_HISTORY_BID"),
            actor: bid.bidder.id,
            pricing: getPricingString(bid.pricing),
            createdAt: bid.createdAtTimestamp,
        }));
        if (((_d = data.pricing.reserve) === null || _d === void 0 ? void 0 : _d.createdAtTimestamp) &&
            (
            // Only show approved auction listings
            (_e = data.pricing.reserve) === null || _e === void 0 ? void 0 : _e.approved)) {
            eventsList.push({
                activityDescription: getString("BID_HISTORY_LISTED"),
                pricing: null,
                actor: data.pricing.reserve.tokenOwner.id,
                // TODO(iain): Update to the timestamp when approved
                createdAt: data.pricing.reserve.createdAtTimestamp,
            });
        }
        if (data.pricing.reserve &&
            data.pricing.reserve.currentBid &&
            data.pricing.status === nft_hooks_1.AuctionStateInfo.RESERVE_AUCTION_ENDED) {
            eventsList.push({
                activityDescription: getString("BID_HISTORY_WON_AUCTION"),
                pricing: null,
                actor: data.pricing.reserve.currentBid.bidder.id,
                createdAt: data.pricing.reserve.expectedEndTimestamp,
            });
        }
        if (data.pricing.reserve &&
            data.pricing.reserve.previousBids.length &&
            data.pricing.status === nft_hooks_1.AuctionStateInfo.RESERVE_AUCTION_FINISHED) {
            eventsList.push({
                activityDescription: getString("BID_HISTORY_WON_AUCTION"),
                pricing: null,
                actor: data.pricing.reserve.previousBids[0].bidder.id,
                createdAt: data.pricing.reserve.finalizedAtTimestamp,
            });
        }
        if ("zoraNFT" in data && data.zoraNFT.createdAtTimestamp) {
            eventsList.push({
                activityDescription: getString("BID_HISTORY_MINTED"),
                pricing: null,
                actor: data.nft.creator || "",
                createdAt: data.zoraNFT.createdAtTimestamp,
            });
        }
        if ("openseaInfo" in data) {
            eventsList.push({
                activityDescription: getString("BID_HISTORY_MINTED"),
                pricing: null,
                actor: data.openseaInfo.creator.address,
                createdAt: null,
            });
        }
        return eventsList
            .sort((bidA, bidB) => (bidA.createdAt > bidB.createdAt ? -1 : 1))
            .map((bidItem) => (react_1.createElement("div", Object.assign({}, getStyles("fullPageHistoryItem"), { key: bidItem.createdAt }),
            jsx_runtime_1.jsxs("div", { children: [jsx_runtime_1.jsxs("span", Object.assign({}, getStyles("pricingAmount"), { children: [jsx_runtime_1.jsx(AddressView_1.AddressView, { address: bidItem.actor }, void 0), " "] }), void 0), bidItem.activityDescription, " ", bidItem.pricing] }, void 0),
            bidItem.createdAt && (jsx_runtime_1.jsx("div", Object.assign({}, getStyles("fullPageHistoryItemDatestamp"), { children: formatDate(bidItem.createdAt) }), void 0)))));
    };
    return (jsx_runtime_1.jsx(InfoContainer_1.InfoContainer, Object.assign({ titleString: "NFT_HISTORY" }, { children: getPastBids() }), void 0));
};
exports.BidHistory = BidHistory;
