// @zoralabs/nft-components/dist/constants/style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Style = void 0;
const css_1 = require("@emotion/css");
const svg_icons_1 = require("./svg-icons");
const theme_1 = require("./theme");
const pricingLayout = (theme) => css_1.css `
  display: grid;
  grid-auto-flow: column;
  grid-template-rows: auto auto;
  grid-auto-column: 1fr;
  padding: ${theme.textBlockPadding};
  border-top: ${theme.borderStyle};
`;
const buttonCommonSize = (size) => `
  padding: ${size};
  width: ${size};
  height: ${size};
  background: #eee;
  border: 0;
  border-radius: 200px;
  color: transparent;
  cursor: pointer;
  background-repeat: no-repeat;
  background-position: center;
  margin: 0 10px;
`;
const CENTER_FULL_CONTAINER = `
  position: absolute;
  width: 100%;
  justify-content: center;
  align-content: center;
  align-items: center;
  justify-items: center;
  display: flex;
  height: 100%;
`;
exports.Style = {
    theme: theme_1.ThemeOptions,
    styles: {
        auctionHouseList: (_) => css_1.css `
      display: flex;
      flex-wrap: wrap;
    `,
        // Styles for preview card
        cardOuter: (theme, { hasClickEvent }) => css_1.css `
      ${hasClickEvent ? "cursor: pointer;" : ""}
      background: ${theme.previewCard.background};
      overflow: hidden;
      border-radius: ${theme.defaultBorderRadius}px;
      border: ${theme.borderStyle};
      margin: 15px;
      width: ${theme.previewCard.width};
      line-height: ${theme.lineSpacing}px;
      ${theme.bodyFont}
      transition: transform 0.1s ease-in-out;
      &:active {
        transform: scale(0.98);
      }
    `,
        cardHeader: (theme) => css_1.css `
      padding: ${theme.textBlockPadding};
      ${theme.titleFont}
    `,
        cardMediaWrapper: (theme) => css_1.css `
      width: ${theme.previewCard.width};
      height: ${theme.previewCard.height};
      display: flex;
      overflow: hidden;
      position: relative;
      justify-content: center;
    `,
        cardItemInfo: (theme) => css_1.css `
      padding: ${theme.textBlockPadding};
      border-top: ${theme.borderStyle};
    `,
        cardAuctionPricing: (theme, { type, }) => {
            const getActiveStyle = () => {
                switch (type) {
                    case "reserve-active":
                        return `
              background: #000;
              color: #fff;
            `;
                    case "reserve-pending":
                        return `
              background: #e6e6e6; 
            `;
                    case "unknown":
                    case "perpetual":
                        return ``;
                }
            };
            return css_1.css `
        display: grid;
        grid-auto-flow: column;
        grid-template-rows: auto auto;
        grid-auto-column: 1fr;
        padding: ${theme.textBlockPadding};
        border-top: ${theme.borderStyle};
        ${getActiveStyle()}
      `;
        },
        cardTitle: (theme) => css_1.css `
      max-width: calc(${theme.previewCard.width} - 30px),
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
      ${theme.titleFont}
    `,
        // Styles for full-page view
        fullPage: (theme) => theme.bodyFont,
        fullMediaWrapper: (_) => css_1.css `
      margin: 5%;
      position: relative;
    `,
        fullItemInfo: (_) => css_1.css ``,
        fullTitle: (_) => css_1.css `
      font-size: 30px;
      margin: 20px 0;
    `,
        fullDescription: (theme) => css_1.css `
      font-size: ${theme.fontSizeFull};
      margin: 10px 0;
    `,
        fullOwnerAddress: (theme) => [
            css_1.css `
        font-size: ${theme.fontSizeFull};
      `,
            theme.titleFont,
        ],
        fullLabel: (theme) => [
            css_1.css `
        text-transform: uppercase;
        font-size: 14px;
        margin-bottom: 5px;
        opacity: 0.5;
      `,
            theme.bodyFont,
        ],
        fullPageHistoryItem: (theme) => [
            css_1.css `
        margin-top: 14px;
        display: flex;
        flex-direction: column;
        font-weight: 300;
      `,
            theme.bodyFont,
        ],
        fullPageHistoryItemDatestamp: (theme) => [
            css_1.css `
        font-size: 12px;
        padding-top: 2px;
        opacity: 0.5;
      `,
            theme.bodyFont,
        ],
        fullPageDataGrid: (_) => css_1.css `
      display: grid;
      grid-gap: 20px;
    `,
        infoContainer: (theme, { bottomPadding }) => css_1.css `
        border: ${theme.borderStyle};
        border-radius: ${theme.defaultBorderRadius}px;
        padding: 20px 20px ${bottomPadding ? "20px" : 0};
        position: relative;
      `,
        fullInfoSpacer: (_, { height = 15 }) => css_1.css `
      height: ${height}px;
    `,
        fullInfoAuctionWrapper: () => ``,
        fullPlaceOfferButton: (_) => css_1.css ``,
        fullInfoCreatorEquityContainer: (_) => css_1.css `
      margin-top: 15px;
    `,
        fullInfoProofAuthenticityContainer: (_) => css_1.css `
      margin-top: 15px;
    `,
        fullProofLink: (theme) => css_1.css `
      display: block;
      text-decoration: none;
      color: ${theme.linkColor};
      padding: 20px;
      margin: 0 -20px;
      border-top: ${theme.borderStyle};

      :hover {
        background-color: #f2f2f2;
      }
      :after {
        content: " ";
        width: 14px;
        height: 14px;
        opacity: 0.5;
        background-image: url("data:image/svg+xml,${encodeURIComponent(svg_icons_1.SVG_NEXT_ICON)}");
        color: #eee;
        right: 20px;
        position: absolute;
      }
    `,
        fullCreatorOwnerSection: (theme) => [
            pricingLayout(theme),
            css_1.css `
        border-top: 0;
      `,
        ],
        // Generic styles
        button: (theme, { primary }) => css_1.css `
      background: ${primary
            ? theme.buttonColor.primaryBackground
            : theme.buttonColor.background};
      color: ${primary
            ? theme.buttonColor.primaryText
            : theme.buttonColor.text};
      border-radius: ${theme.defaultBorderRadius}px;
      padding: 11px;
      font: inherit;
      text-decoration: none;
      margin: 0;
      border: 0;
      cursor: pointer;
      display: inline-block;
      transition: transform 0.1s ease-in-out;
      &:active {
        transform: scale(0.98);
      }
    `,
        textSubdued: (theme) => [
            css_1.css `
        opacity: 0.5;
      `,
            theme.bodyFont,
        ],
        pricingAmount: (theme) => theme.titleFont,
        mediaLoader: (_, { mediaLoaded, isFullPage }) => css_1.css `
      pointer-events: none;
      ${isFullPage ? "min-height: 40vh;" : ""}
      width: 100%;
      justify-content: center;
      align-items: center;
      opacity: ${mediaLoaded ? "0" : "1"};
      transition: 0.2s ease-out opacity;
      ${CENTER_FULL_CONTAINER}
      ${isFullPage && !mediaLoaded
            ? `
      &:after {
        content: " ";
        ${isFullPage ? "height: 30vh" : ""}
      }
      `
            : ""}
    `,
        mediaObject: (_, { mediaLoaded, isFullPage }) => css_1.css `
      opacity: ${mediaLoaded ? "1" : "0"};
      transition: 0.2s ease-in opacity;
      ${isFullPage ? "max-height: 70vh;" : ""}
      ${isFullPage ? "max-width: 100%;" : ""}
      display: block;
      margin: 0 auto;
      flex-shrink: 1;
    `,
        mediaAudioWrapper: (_) => css_1.css `
      margin-top: 40px;
      display: flex;
      justify-content: center;
      align-items: center;
      flex-direction: column;
      width: 100%;
    `,
        mediaAudioWaveform: (_) => css_1.css `
      width: 100%;
      cursor: pointer;
    `,
        mediaObjectMessage: (_) => css_1.css `
      align-self: center;
    `,
        mediaContentText: (theme) => [
            css_1.css `
        white-space: pre;
        text-align: left;
        padding: 20px;
        width: 100%;
      `,
            theme.mediaContentFont,
        ],
        mediaPlayButton: (_, { playing }) => css_1.css `
      ${buttonCommonSize("32px")}
      background-image: url("data:image/svg+xml,${encodeURIComponent(playing ? svg_icons_1.SVG_PAUSE : svg_icons_1.SVG_PLAY_ARROW)}");
    `,
        mediaVideoControls: (_, { isFullPage }) => css_1.css `
      ${CENTER_FULL_CONTAINER}
      ${isFullPage ? `
        opacity: 0;
        &:hover {
          opacity: 1;
        }
        transition: opacity 0.6s ease-in-out;
        transition-delay: 0 0.3s;
      ` : "display: none;"}
    `,
        mediaFullscreenButton: (_) => css_1.css `
      ${buttonCommonSize("16px")}
      background-color: #000;
      background-image: url("data:image/svg+xml,${encodeURIComponent(svg_icons_1.SVG_FULLSCREEN)}");
    `,
        mediaMuteButton: (_, { muted }) => css_1.css `
      ${buttonCommonSize("16px")}
      background-color: #000;
      background-image: url("data:image/svg+xml,${encodeURIComponent(muted ? svg_icons_1.SVG_UNMUTED : svg_icons_1.SVG_MUTED)}");
    `,
    },
};
